messaging/src/main/java/org/axonframework/eventhandling/gateway/AbstractEventGateway.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2010-2019. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventhandling.gateway;
+
+import org.axonframework.common.AxonConfigurationException;
+import org.axonframework.common.Registration;
+import org.axonframework.eventhandling.EventBus;
+import org.axonframework.eventhandling.EventMessage;
+import org.axonframework.messaging.MessageDispatchInterceptor;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import static java.util.Arrays.asList;
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+import static org.axonframework.eventhandling.GenericEventMessage.asEventMessage;
+
+/**
+ * Abstract implementation of an EventGateway, which handles the dispatch interceptors. The
+ * actual publishing of events is left to the subclasses.
+ *
+ * @author Bert Laverman
+ * @since 4.0.4
+ */
+public abstract class AbstractEventGateway {
```

Comment:
You could've done without an `AbstractEventGateway` I think.

The only reason there's an `AbstractCommandGateway` is because we have several implementations of the `CommandGateway` interface.

This is however the introduction of the first `EventGateway`, so this abstract level might not add that much.

Reply:
So blame me for being an architect and thinking ahead. Want me to remove it?

Reply:
Nah, I think this is good. That comment was just an FYI :+1: